bigtable/client/detail/bulk_mutator_test.cc

```diff
@@ -0,0 +1,379 @@
+    EXPECT_TRUE(status.ok());
+  }
+  auto failures = mutator.ExtractFinalFailures();
+  EXPECT_TRUE(failures.empty());
+}
+
+/// @test Verify that MultipleRowsMutator only retries idempotent mutations
+TEST(MultipleRowsMutatorTest, RetryOnlyIdempotent) {
+  namespace btproto = ::google::bigtable::v2;
+  namespace bt = bigtable;
+  using namespace ::testing;
+
+  // Create a BulkMutation with a non-idempotent mutation.
+  bt::BulkMutation mut(
+      bt::SingleRowMutation("foo", {bigtable::SetCell("fam", "col", "baz")}),
+      bt::SingleRowMutation("bar", {bigtable::SetCell("fam", "col", 0, "qux")}),
+      bt::SingleRowMutation("baz", {bigtable::SetCell("fam", "col", "v")}));
+
+  // We will setup the mock to return recoverable failures for idempotent
+  // mutations.
+  auto r1 = absl::make_unique<MockReader>();
+  EXPECT_CALL(*r1, Read(_))
+      .WillOnce(Invoke([](btproto::MutateRowsResponse* r) {
+        // ... simulate a partial (recoverable) failure ...
+        auto& e0 = *r->add_entries();
+        e0.set_index(0);
+        e0.mutable_status()->set_code(grpc::UNAVAILABLE);
+        auto& e1 = *r->add_entries();
+        e1.set_index(1);
+        e1.mutable_status()->set_code(grpc::UNAVAILABLE);
+        return true;
+      }))
+      .WillOnce(Return(false));
+  EXPECT_CALL(*r1, Finish()).WillOnce(Return(grpc::Status::OK));
+
+  // The BulkMutator should issue a second request, with only the idempotent
+  // mutations, make the mocks return success for them.
+  auto r2 = absl::make_unique<MockReader>();
+  EXPECT_CALL(*r2, Read(_))
+      .WillOnce(Invoke([](btproto::MutateRowsResponse* r) {
+        {
+          auto& e = *r->add_entries();
+          e.set_index(0);
+          e.mutable_status()->set_code(grpc::OK);
+        }
+        return true;
+      }))
+      .WillOnce(Return(false));
+  EXPECT_CALL(*r2, Finish()).WillOnce(Return(grpc::Status::OK));
+
+  // Verify that the second response has the right contents.  It is easier (and
+  // more readable) to write these in a separate small lambda expression because
+  // ASSERT_EQ() has an embedded "return;" in it, which does not play well with
+  // the rest of the stuff here.
+  auto expect_r2 = [](btproto::MutateRowsRequest const& r) {
+    ASSERT_EQ(1, r.entries_size());
+    EXPECT_EQ("bar", r.entries(0).row_key());
+  };
+  btproto::MockBigtableStub stub;
+  EXPECT_CALL(stub, MutateRowsRaw(_, _))
+      .WillOnce(Invoke(
+          [&r1](grpc::ClientContext*, btproto::MutateRowsRequest const& r) {
+            EXPECT_EQ(3, r.entries_size());
+            return r1.release();
+          }))
+      .WillOnce(Invoke([&r2, expect_r2](grpc::ClientContext*,
+                                        btproto::MutateRowsRequest const& r) {
+        expect_r2(r);
+        return r2.release();
+      }));
+
+  auto policy = bt::DefaultIdempotentMutationPolicy();
+  bt::detail::BulkMutator mutator("foo/bar/baz/table", *policy, std::move(mut));
+
+  for (int i = 0; i != 2; ++i) {
```

Comment:
Add a comment here that 2 is because of `reader1` and `reader2`?

Reply:
Done.